---
title: "Run UMAP and Clustering"
author: "Brennan McDonald"
date: "2024-04-02"
output: html_document
---

# Overview

The following script contains the code used to run the UMAP and clustering algorithms on the Seurat object containing the normalized data for all 12 timepoints in this study. It also contains code for identifying the marker genes for each cluster, assigning cell type identities to clusters, and generating several plots used in the main paper.


```{r}
# Load packages
library(tidyverse)
library(Seurat)
library(scCustomize)
library(viridis)
```

```{r}
#setwd("/Users/brennandmcdonald/Documents/Wray Lab/scRNA-seq Project/HE scRNA-seq")
CreateSeuratObject(load(
  file = "HE_6-60hpf_integrated-SCT6k_195dim.Rda", 
                        verbose = T))
```


# Obtain summary statistics for the dataset

```{r}
times <- str_extract(he.integrated@meta.data$Stage, "\\d+")
times <- as.integer(times)

he.integrated <- AddMetaData(he.integrated, times, col.name = "time_int")

nrow(he.integrated@meta.data)

# Creates a violin plot of the distribution of number of RNA features and RNA 
# transcripts found per cell in the dataset.
# See figure S____
VlnPlot(he.integrated, features = c("nFeature_RNA", "nCount_RNA"),
        group.by = "time_int", pt.size = 0)

# Outputs the average number of RNA features and transcripts over cells in each
# timepoint.
he.integrated@meta.data %>% 
  group_by(time_int) %>% 
  summarize(average_nFeature = mean(nFeature_RNA),
            average_nCount = mean(nCount_RNA))
```


# Run the UMAP and clustering algorithms

```{r}
# Perform clustering with resolution = 3, which returns 57 distinct clusters
he.integrated <- FindNeighbors(he.integrated, dims = 1:195)
he.integrated <- FindClusters(he.integrated, resolution = 3)

# Non-linear dimensional reduction (UMAP)
he.integrated <- RunUMAP(he.integrated, dims = 1:195)
```


# Identify the marker genes for each cluster

```{r}
# Find cluster marker genes that are expressed in a minimum of 25% of cells in each cluster with a log2FC value of >0.25
he.integrated.markers <- FindAllMarkers(he.integrated, only.pos = T, min.pct = 0.25, 
                                        logfc.threshold = 0.25)

he.markers.save <- he.integrated.markers %>% 
  group_by(cluster) %>% 
  slice_max(n = 50, order_by = avg_log2FC)

write.csv(he.markers.save, file = "data-files/HE_6-60hr_cluster-markers.csv")
```


Use the marker genes and previous literature to assign cell type identities to each of the clusters (see table S___ for the genes used to make each assignment).

```{r}
Idents(he.integrated) <- "seurat_clusters"


# The index of these celltype identities in the vector corresponds to the 
# number of the cluster that the identity will be assigned to
new.cluster.ids <- c("Animal Pole Domain",
                     "Animal Pole Domain",
                     "Larval Ectoderm",
                     "Larval Ectoderm",
                     "Larval Ectoderm",
                     "Pigment",
                     "Larval Ectoderm",
                     "Ciliary Band",
                     "Early Ectoderm",
                     "Larval Ectoderm",
                     "Animal Pole Domain",
                     "Ciliary Band",
                     "Vegetal Ectoderm",
                     "Larval Ectoderm",
                     "Larval Ectoderm",
                     "Ciliary Band",
                     "Larval Ectoderm",
                     "Early Endomesoderm",
                     "Rudiment Endomesoderm",
                     "Gut",
                     "Left Coelomic Pouch",
                     "Animal Pole Domain",
                     "Blastocoelar",
                     "Vegetal Ectoderm",
                     "Vestibular Ectoderm",
                     "Vestibular Ectoderm",
                     "Blastocoelar",
                     "Left Coelomic Pouch",
                     "Early Endomesoderm",
                     "Pigment",
                     "Animal Pole Domain",
                     "Neural",
                     "Rudiment Endomesoderm",
                     "Vegetal Ectoderm",
                     "Larval Ectoderm",
                     "Unknown",
                     "Vegetal Ectoderm",
                     "Ciliary Band",
                     "Early Mesoderm",
                     "Neural",
                     "Undifferentiated",
                     "Right Coelomic Pouch",
                     "Animal Pole Domain",
                     "Left Coelomic Pouch",
                     "Pigment",
                     "Left Coelomic Pouch",
                     "Larval Ectoderm",
                     "Rudiment Endomesoderm",
                     "Early Endomesoderm",
                     "Skeletal",
                     "Vestibular Ectoderm",
                     "Neural",
                     "Unknown",
                     "Neural",
                     "Neural",
                     "Unknown",
                     "Neural")

# The identities are added to a new identity class in the Seurat object
names(new.cluster.ids) <- levels(he.integrated)
he.integrated <- RenameIdents(he.integrated, new.cluster.ids)

# The new identities are added to the metadata of the Seurat object
he.integrated <- AddMetaData(
  object = he.integrated,
  metadata = Idents(he.integrated),
  col.name = "celltype"
)

# The updated Seurat object is saved
save(he.integrated, file = "HE_6-60hpf_integrated-SCT6k.Rda")
```


The following code uses the scCustomize package to generate DimPlots for the UMAP projection of the data colored by sample timepoint and celltype identity

```{r}
# Generate a list of sample timepoints (the number values only)
times <- str_extract(he.integrated@meta.data$Stage, "\\d+")
times <- as.integer(times)

he.integrated <- AddMetaData(he.integrated, times, col.name = "time_int")

x <- seq(10,250,20)
viridis_plasma_dark_high[x]

# This generates the UMAP colored by time used in Figure 2
DimPlot_scCustom(he.integrated,
                 group.by = "time_int",
                 colors_use = viridis_plasma_dark_high[x]) +
  theme(legend.position = "none")

# This generates the UMAP colored by celltype used in Figure 2
DimPlot_scCustom(he.integrated,
                 group.by = "celltype",
                 colors_use = "alphabet") +
  theme(legend.position = "none")
```


The following code extracts the marker genes for each celltype in the Seurat object.

```{r}
# Get cell type markers
celltype.markers <- FindAllMarkers(he.integrated, only.pos = T, min.pct = 0.25, 
                                        logfc.threshold = 0.25)

top10markers <- celltype.markers %>%
  group_by(cluster) %>%
  slice_max(n = 50, order_by = desc(pct.2))

write.csv(top10markers, file = "data-files/HE_6-60hr_celltype-markers.csv")
```

The following code uses the curated set of celltype marker genes (from table S___) to generate a dotplot showing expression of each of the genes in each celltype for Figure S___.

```{r}
library(scCustomize)

all_celltype_genes <- c("HER-17729.t1:Sp-Ccnr2",
"HER-33043.t1:Sp-Yp303",
"HER-8313.t1:Sp-El1L4",
"HER-48315.t1:Sp-C8orf48L",
"HER-16499.t1:Sp-ClvhH1",
"HER-28477.t1:Sp-Guan/Cyc",
"HER-25973.t1:Sp-Igf1",
"HER-10974.t1:Sp-Hbn",
"HER-17866.t1:Sp-Six3",
"HER-16550.t1:Sp-FoxQ2-1",
"HER-23275.t1:Sp-Fzd5/8",
"HER-20226.t1:Sp-SoxB2",
"HER-17453.t1:Sp-Univin",
"HER-10887.t1:Sp-FoxABL",
"HER-24103.t1:Sp-Nodal",
"HER-18058.t1:Sp-Gsc",
"HER-348.t1:Sp-Tbx2/3",
"HER-17677.t1:Sp-FoxA",
"HER-40064.t1:Sp-Hox11/13b",
"HER-17599.t1:Sp-FoxN2/3",
"HER-21330.t1:Sp-Wnt1",
"HER-5916.t1:Sp-Scl",
"HER-38449.t1:Sp-GataC",
"HER-12631.t1:Sp-Ese",
"HER-9027.t1:Sp-Hhex",
"HER-25978.t1:Sp-Endo16",
"HER-17855.t1:Sp-Six1/2",
"HER-29126.t1:Sp-HesA",
"HER-6581.t1:Sp-FoxD-1",
"HER-46925.t1:Sp-Mif4",
"HER-46918.t1:Sp-Pks1",
"HER-21404.t1:Sp-Hox11/13c",
"HER-47443.t1:Sp-Fmo3",
"HER-40156.t1:Sp-185/333D1",
"HER-45847.t1:Sp-Astacin4",
"HER-44751.t1:Sp-Irf4",
"HER-29337.t1:Sp-Msp130r2",
"HER-12794.t1.2.5f2c6949:Sp-Alx1",
"HER-26255.t1:Sp-Sm37",
"HER-40062.t1:Sp-Hox8-1",
"HER-21329.t1:Sp-Wnt9",
"HER-23589.t1:Sp-Pitx2",
"HER-39665.t1:Sp-Dlx",
"HER-16107.t1:Sp-SoxE",
"HER-23234.t1:Sp-Prdm14-1",
"HER-17538.t1:Sp-CalmL1",
"HER-45795.t1:Sp-Secrtag",
"HER-10174.t1:Sp-Syt4",
"HER-17334.t1:Sp-Giant",
"HER-13245.t1-HER-13243.t1:Sp-B4galt6-3",
"HER-26992.t1:Sp-Acsc",
"HER-39411.t1:Sp-Smadip",
"HER-16452.t1-HER-16451.t1:Sp-Egfi",
"HER-9194.t1:Sp-Msx",
"HER-21545.t1:Sp-Wnt3",
"HER-29039.t1:Sp-Fn3/Egf-2",
"HER-33009.t1:Sp-Onecut",
"HER-45860.t1:Sp-Btub3",
"HER-35876.t1:Sp-Atub13",
"HER-28.t1:Sp-FbsL-2",
"HER-40059.t1:Sp-Hox7",
"HER-10395.t1-HER-10396.t1:Sp-WntA",
"HER-48747.t1:Sp-Lox",
"HER-45097.t1:Sp-CebpaL",
"HER-47859.t1:Sp-Hsp701D",
"HER-17148.t1:Sp-GlxnL",
"HER-33417.t1:Sp-Rfwd3",
"HER-36368.t1:Sp-Ocrl1")

levs <- factor(c("Early Ectoderm",
                 "Undifferentiated",
                 "Animal Pole Domain",
                 "Larval Ectoderm",
                 "Early Endomesoderm",
                 "Early Mesoderm",
                 "Gut",
                 "Pigment",
                 "Blastocoelar",
                 "Skeletal",
                 "Right Coelomic Pouch",
                 "Left Coelomic Pouch",
                 "Neural",
                 "Rudiment Endomesoderm",
                 "Vestibular Ectoderm",
                 "Ciliary Band",
                 "Vegetal Ectoderm",
                 "Unknown"))

Idents(he.integrated) <- factor(he.integrated@active.ident, levels = levs)

# This generates the dotplot of celltype marker genes for Figure S___
DotPlot_scCustom(he.integrated,
                 features = all_celltype_genes,
                 x_lab_rotate = T,
                 flip_axes = T)
```



## Export featureplots for figure 8

```{r}
library(scCustomize)
library(viridis)

search <- "Msp130r2"
gene_id <- rownames(he.integrated@assays$RNA@counts)[grep(search, rownames(he.integrated@assays$RNA@counts))]
gene_id

#AverageExpression(he.integrated, features = gene_id[1])
#AggregateExpression(he.integrated, features = gene_id[1])
FeaturePlot(he.integrated, features = gene_id[1])

# Export as 6x8.5 for figure 8
FeaturePlot_scCustom(he.integrated, 
                     features = gene_id[1],
                     colors_use = viridis_dark_high,
                     pt.size = 0.3,
                     order = T # need to decide on this
                     )
```










