---
title: "InterProScan Output Processing"
author: "Brennan McDonald"
date: "2024-04-03"
output: html_document
---

# Overview

This script provides the code needed to convert the gene function annotation output (of the protein-coding genes in the He genome) from InterProScan into an orgDb package, using the tools provided by the `AnnotationForge` package. This will make it easy to filter the gene function annotation output for GO terms of interest and to perform GO over representation analyses.

```{r}
# Load packages
library(tidyverse)
library(AnnotationForge)
```


# Make an orgdb package for He

The following code will construct an orgDb package for He, using the InterProScan results as input. Unfortunately, the InterProScan output files are not in the GAF format required for input into the orgDb pipeline. I used the functionality provided by the Cyverse online platform to convert the output .xml file from InterProScan into the GAF format I needed. This involved using the InterProScan Results Function script described here: https://cyverse.atlassian.net/wiki/spaces/DEapps/pages/241881762/InterProScan+Results+Function.

```{r}
# Import the InterProScan GO annotations, in GAF format
go <- read.delim("data-files/interproscan-output_Hery_peptide_models.fasta_gaf.txt")

# Import the He genome annotations from Davidson et al. (2022)
anno <- read.delim("data-files/H_ery_SPU_hits_best.txt")
```


The GO annotation and He genome ananotation dataframes need to be modified into the correct format for input into the orgDb pipeline.

```{r}
# Modify dataframes for input. 
# A `GID` (gene ID) column needs to be the same across all of them
go <- go %>% 
  mutate(GID = DB_Object_ID, .before = Database)

anno <- anno %>% 
  mutate(GID = H_ery_ID, .before = H_ery_ID) %>% 
  rename(SPU_hit = SPU.hit, SPU_pident = SPU.pident, SPU_evalue = SPU.evalue,
         SPU_common_name = SPU.common_name, SPU_family = SPU.family,
         SPU_description = SPU.description)

anno$SPU_common_name <- replace(anno$SPU_common_name, is.na(anno$SPU_common_name), "NA")
anno$SPU_family <- replace(anno$SPU_family, is.na(anno$SPU_family), "NA")
anno$SPU_description <- replace(anno$SPU_description, is.na(anno$SPU_description), "NA")

# Only the GID and GO_ID columns are retained from the GO dataframe for downstream
# steps.
go_extra <- go %>% 
  select(GID, GO_ID, Evidence_Code) %>% 
  rename(GO = GO_ID, EVIDENCE = Evidence_Code)
```

The `makeOrgPackage` function from `AnnotationForge` is used to make the orgDb package for He. This is saved to an output directory, and can be installed as a package in R called `org.Herythrogramma.eg.db`.

```{r}
makeOrgPackage(gene_info = anno, 
               go = go_extra, 
               version = "1.0.0",
               maintainer = "[NAME] <EMAIL>",
               author = "[NAME] <EMAIL>",
               outputDir = "data-files/",
               tax_id = "7634", # taxon ID for He from NCBI
               genus = "Heliocidaris",
               species = "erythrogramma",
               goTable = "go",
               verbose = T)

install.packages("data-files/org.Herythrogramma.eg.db", repos = NULL)
```

Test loading the package to verify that the installation was successful.

```{r}
library(org.Herythrogramma.eg.db)
```





