---
title: "Curated TF List"
author: "Brennan McDonald"
date: "2024-04-03"
output: html_document
---

# Overview

This script provides the code needed to generate the list of candidate transcription factors found in the He genome, based on GO term annotations and a previously used list.

```{r}
# Load packages
library(tidyverse)
library(org.Herythrogramma.eg.db) # This package was created in the "interproscan-output-processing.Rmd" script and provides the GO annotations
library(clusterProfiler)
```


# TF list curation

I first used the GO annotations I generated for all genes in the He genome using InterProScan. I filter gene GO annotations for GO IDs corresponding to, or below "Transcription regulator activity" in the GO Molecular Function hierarchy (provided by the AmiGO database, https://amigo.geneontology.org/amigo).

GO terms to use (all under Molecular Function):
- GO:0003700 - DNA-binding transcription factor activity
- GO:0001216 - DNA-binding transcription activator activity
- GO:0001227 - DNA-binding transcription repressor activity, RNA polymerase II-specific
- GO:0001228 - DNA-binding transcription activator activity, RNA polymerase II-specific
- GO:0000981 - DNA-binding transcription factor activity, RNA polymerase II-specific
- GO:0001217 - DNA-binding transcription repressor activity

The functionality provided by the `AnnotationDbi` package is used to filter the `org.Herythrogramma.eg.db` (created using the "interproscan-output-processing.Rmd" script) database for the GO terms associated with He genes.

```{r}
k <- keys(org.Herythrogramma.eg.db, keytype = "SPU_common_name")

# Filter the gene list for genes with GO terms that have molecular functions related to DNA binding transcriptional regulation
tfs <- AnnotationDbi::select(org.Herythrogramma.eg.db, keys = k, 
                      columns = c("GO", "SPU_description", "GID", "SPU_hit"),
                      keytype = "SPU_common_name") %>% 
  dplyr::filter(GO %in% c("GO:0003700", "GO:0001216", "GO:0001227", "GO:0001228", "GO:0000981", "GO:0001217")) %>% 
  distinct(SPU_common_name, .keep_all = T)
```


I generated a separate list of TFs using the gene list analyzed in Devens et al. (2023). This list was originally retrieved from an old version of Echinobase (https://www.echinobase.org/echinobase/). I filtered this list of genes for genes that were either labeled as "TranscriptionFactor" or "ZincFinger", both of which have DNA-binding transcription regulatory abilities.

```{r}
# list of Echinobase genes from Devens et al. (2023)
devens <- read.csv("data-files/functional_category_table.csv", row.names = NULL)

# filter list for SPU_hits listed as TranscriptionFactors or ZincFingers
devens_tf <- devens %>% 
  filter(Function == "TranscriptionFactor")

devens_zf <- devens %>% 
  filter(Function == "ZincFinger")

# Collect the SPU_hits
devens_tf_keys <- devens_tf$Gene
devens_zf_keys <- devens_zf$Gene
```


I then merged the two lists, such that there were no repeated SPU gene IDs. This was saved in a file for downstream analyses of TF expression patterns.

```{r}
# Merge the SPU hits from the GO term list and Hannah's list into union sets
un_tf <- union(devens_tf_keys, tfs$SPU_hit)
un_zf <- union(devens_zf_keys, tfs$SPU_hit)
un_all <- union(un_tf, un_zf)

# Create a full list of the genes using both the GO terms and the list from
# Devens et al. (2023)
devens_go_tfs <- AnnotationDbi::select(org.Herythrogramma.eg.db, keys = un_all, 
                      columns = c("SPU_common_name", "GO", "SPU_description", "GID", "SPU_hit"),
                      keytype = "SPU_hit") %>% 
  distinct(GID, .keep_all = T) %>% # filter for distinct HER transcripts
  filter(!is.na(GID))


write.csv(devens_go_tfs, "data-files/devens-and-go-term_tf-list.csv",
          row.names = F)
```



