---
title: "Skeletogenic Cell Analysis"
author: "Brennan McDonald"
date: "2024-04-03"
output: html_document
---

# Overview

The following script provides the ability to run the analyses of the skeletogenic cell lineage in the scRNA-seq dataset. The script contains the code needed to generate the plots used in Figures 3 and 4.


```{r}
# Load packages
library(tidyverse)
library(Seurat)
library(scCustomize)
```

```{r}
# Load the He Seurat object
CreateSeuratObject(load(
  file = "HE_6-60hpf_integrated-SCT6k_195dim.Rda", 
                        verbose = T))
```

The Seurat object for Lytechinus variegatus was retrieved from 

```{r}
lv.integrated <- readRDS(".rds")
```


This is the code that produces the DimPlot highlighting the skeletogenic cell cluster in the UMAP for He (Figure 3).

```{r}
Cluster_Highlight_Plot(seurat_object = he.integrated,
                       cluster_name = "Skeletal",
                       highlight_color = "red",
                       pt.size = 0.3) +
  theme(legend.position = "none")
```


This code produces the cluster highlight plot for the skeletogenic cell lineage in Lv, as well as the overall UMAP for Lv colored by timepoint (see Figure 3).

```{r}
# Produce the UMAP plot highlighting the skeletogenic cell lineage for Lv (Figure 3)
Cluster_Highlight_Plot(lv.integrated,
                       cluster_name = c("PMCs", "Micromeres"),
                       highlight_color = c("orange", "orange"),
                       pt.size = 0.3) +
  theme(legend.position = "none")

lv.integrated@meta.data %>% View()


times <- str_extract(lv.integrated@meta.data$Stage, "\\d+")
times <- as.integer(times)

unique(times)

lv.integrated <- AddMetaData(lv.integrated, times, col.name = "time_int")

x <- seq(10,250,14)
viridis_plasma_dark_high[x]

# Produce the UMAP plot colored by sample time point for Figure 3
DimPlot_scCustom(lv.integrated,
                 group.by = "time_int",
                 colors_use = viridis_plasma_dark_high[x]) +
  theme(legend.position = "none")
```



# Co-expression analysis

## Custom plotting function

This code takes a vector of cell UMIs that are known to co-express the two genes at a low or high level, and then produces a ggplot2 version of the UMAP with those cells highlighted.

```{r}
coexp_plot <- function(coords, low_exp, high_exp) {
  coords1 <- coords %>% 
    mutate(Expression = case_when(Cell %in% low_exp ~ "Low Coexpression",
                                  .default = "No Coexpression")) %>% 
    mutate(Expression = if_else(Cell %in% high_exp, "High Coexpression", Expression))
    
  coords1$Expression <- factor(coords1$Expression, 
         levels = c("No Coexpression", "Low Coexpression", "High Coexpression"))
  
  coords1 %>%   
    arrange(Expression) %>% 
    ggplot(mapping = aes(x = UMAP_1, y = UMAP_2, 
                         color = Expression)) +
    geom_point(size = 0.3) +
    scale_color_discrete(type = c("gray", "lightgreen", "darkgreen")) +
    labs(color = "Coexpression Level") +
    theme_classic() +
    theme(legend.position = "none")
}
```

## Co-expression analysis for He

### Generate co-expression plots for Figures 3 and 4

Obtain the UMAP coordinates for each of the cells in the He Seurat object

```{r}
umap_coords <- as_tibble(Embeddings(he.integrated, reduction = "umap"))
umap_coords$Cell <- rownames(Embeddings(he.integrated, reduction = "umap"))
```

Generate a co-expression plot for Alx1 and Scl for He. Note that Alx1 is split into two transcripts in the He RNA-seq transcript annotations.

```{r}
# Calculate low and high co-expressed cells for the first Alx1 transcript
alx_scl1_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-5916.t1:Sp-Scl` >= 0.5)
alx_scl1_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-5916.t1:Sp-Scl` >= 1)

# Calculate low and high co-expressed cells for the second Alx1 transcript
alx_scl2_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-5916.t1:Sp-Scl` >= 0.5)
alx_scl2_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-5916.t1:Sp-Scl` >= 1)

# Merge the co-expressing cells for the two transcripts
alx_scl_t05_merge <- union(alx_scl1_t05, alx_scl2_t05)
alx_scl_t1_merge <- union(alx_scl1_t1, alx_scl2_t1)

# Co-expression plot for Alx1 and Scl in Figure 3
coexp_plot(coords = umap_coords, 
           low_exp = alx_scl_t05_merge, 
           high_exp = alx_scl_t1_merge)
```


Generate a co-expression plot for Alx1 and Ese for He. Note that Alx1 is split into two transcripts in the He RNA-seq transcript annotations.

```{r}
alx_ese1_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12631.t1:Sp-Ese` >= 0.5)
alx_ese1_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-12631.t1:Sp-Ese` >= 1)

alx_ese2_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12631.t1:Sp-Ese` >= 0.5)
alx_ese2_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-12631.t1:Sp-Ese` >= 1)

alx_ese_t05_merge <- union(alx_ese1_t05, alx_ese2_t05)
alx_ese_t1_merge <- alx_ese2_t1 # No high co-expressing cells were found for the first Alx1 transcript

coexp_plot(coords = umap_coords, 
           low_exp = alx_ese_t05_merge, 
           high_exp = alx_ese_t1_merge)
```

Generate a co-expression plot for Alx1 and GataC for He. Note that Alx1 is split into two transcripts in the He RNA-seq transcript annotations.

```{r}
alx_gc1_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-38449.t1:Sp-GataC` >= 0.5)
alx_gc1_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-38449.t1:Sp-GataC` >= 1)

alx_gc2_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-38449.t1:Sp-GataC` >= 0.5)
alx_gc2_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-38449.t1:Sp-GataC` >= 1)

alx_gc_t05_merge <- union(alx_gc1_t05, alx_gc2_t05)
alx_gc_t1_merge <- union(alx_gc1_t1, alx_gc2_t1)


coexp_plot(coords = umap_coords, 
           low_exp = alx_gc_t05_merge, 
           high_exp = alx_gc_t1_merge)
```

Generate a co-expression plot for Alx1 and Tbr for He. Note that Alx1 is split into two transcripts in the He RNA-seq transcript annotations.

```{r}
alx_tbr1_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-21600.t1:Sp-Tbr` >= 0.5)
alx_tbr1_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-21600.t1:Sp-Tbr` >= 1)

alx_tbr2_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-21600.t1:Sp-Tbr` >= 0.5)
alx_tbr2_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-21600.t1:Sp-Tbr` >= 1)

alx_tbr_t05_merge <- union(alx_tbr1_t05, alx_tbr2_t05)
alx_tbr_t1_merge <- c() # No high co-expressing cells were found for either Alx1 transcript


coexp_plot(coords = umap_coords, 
           low_exp = alx_tbr_t05_merge, 
           high_exp = alx_tbr_t1_merge)
```


Generate a co-expression plot for Alx1 and FoxO for He. Note that Alx1 is split into two transcripts in the He RNA-seq transcript annotations.

```{r}
alx_foxo1_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-35682.t1:Sp-FoxO-1` >= 0.5)
alx_foxo1_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-35682.t1:Sp-FoxO-1` >= 1)

alx_foxo2_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-35682.t1:Sp-FoxO-1` >= 0.5)
alx_foxo2_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-35682.t1:Sp-FoxO-1` >= 1)

alx_foxo_t05_merge <- union(alx_foxo1_t05, alx_foxo2_t05)
alx_foxo_t1_merge <- c(alx_foxo2_t1) # No high co-expressing cells were found for the first ALx1 transcript

coexp_plot(coords = umap_coords, 
           low_exp = alx_foxo_t05_merge, 
           high_exp = alx_foxo_t1_merge)
```


Generate a co-expression plot for Alx1 and FoxB for He. Note that Alx1 is split into two transcripts in the He RNA-seq transcript annotations.

```{r}
alx_foxb1_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-11066.t1:Sp-FoxB` >= 0.5)
alx_foxb1_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-11066.t1:Sp-FoxB` >= 1)

alx_foxb2_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-11066.t1:Sp-FoxB` >= 0.5)
alx_foxb2_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-11066.t1:Sp-FoxB` >= 1)

alx_foxb_t05_merge <- union(alx_foxb1_t05, alx_foxb2_t05)
alx_foxb_t1_merge <- c() # No high co-expressing cells were found for either Alx1 transcript


coexp_plot(coords = umap_coords, 
           low_exp = alx_foxb_t05_merge, 
           high_exp = alx_foxb_t1_merge)
```


Generate a co-expression plot for Alx1 and Tel for He. Note that Alx1 is split into two transcripts in the He RNA-seq transcript annotations.

```{r}
alx_tel1_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12611.t1:Sp-Tel` >= 0.5)
alx_tel1_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-12611.t1:Sp-Tel` >= 1)

alx_tel2_t05 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12611.t1:Sp-Tel` >= 0.5)
alx_tel2_t1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 1 & 
                        `HER-12611.t1:Sp-Tel` >= 1)

alx_tel_t05_merge <- union(alx_tel1_t05, alx_tel2_t05)
alx_tel_t1_merge <- union(alx_tel1_t1, alx_tel2_t1)


coexp_plot(coords = umap_coords, 
           low_exp = alx_tel_t05_merge, 
           high_exp = alx_tel_t1_merge)
```


### Count numbers of coexpressing cells in the He skeletal lineage

The following code counts the total number of cells that express either Alx1 transcript in the He skeletogenic cell cluster.

```{r}
# number of Alx1 expressing cells

alx_1 <- WhichCells(he.integrated, 
                       expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5,
                       idents = "Skeletal")
alx_2 <- WhichCells(he.integrated, 
                       expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5,
                       idents = "Skeletal")

nalx <- length(union(alx_1, alx_2))
```


The following code counts the number of co-expressing cells for Alx1 and Scl (both high and low co-expression) in the He skeletogenic cell cluster.

```{r}
alx_scl_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-5916.t1:Sp-Scl` >= 0.5,
                       idents = "Skeletal")
alx_scl_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-5916.t1:Sp-Scl` >= 0.5,
                       idents = "Skeletal")

nalx_scl <- length(union(alx_scl_1, alx_scl_2))
```


The following code counts the number of co-expressing cells for Alx1 and Ese (both high and low co-expression) in the He skeletogenic cell cluster.

```{r}
alx_ese_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12631.t1:Sp-Ese` >= 0.5,
                       idents = "Skeletal")
alx_ese_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12631.t1:Sp-Ese` >= 0.5,
                       idents = "Skeletal")

nalx_ese <- length(union(alx_ese_1, alx_ese_2))
```


The following code counts the number of co-expressing cells for Alx1 and GataC (both high and low co-expression) in the He skeletogenic cell cluster.

```{r}
alx_gc_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-38449.t1:Sp-GataC` >= 0.5,
                       idents = "Skeletal")
alx_gc_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-38449.t1:Sp-GataC` >= 0.5,
                       idents = "Skeletal")

nalx_gc <- length(union(alx_gc_1, alx_gc_2))
```


The following code counts the number of co-expressing cells for Alx1 and Tbr (both high and low co-expression) in the He skeletogenic cell cluster.

```{r}
alx_tbr_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-21600.t1:Sp-Tbr` >= 0.5,
                       idents = "Skeletal")
alx_tbr_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-21600.t1:Sp-Tbr` >= 0.5,
                       idents = "Skeletal")

nalx_tbr <- length(union(alx_tbr_1, alx_tbr_2))
```


The following code counts the number of co-expressing cells for Alx1 and FoxO (both high and low co-expression) in the He skeletogenic cell cluster.

```{r}
alx_foxo_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-35682.t1:Sp-FoxO-1` >= 0.5,
                       idents = "Skeletal")
alx_foxo_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-35682.t1:Sp-FoxO-1` >= 0.5,
                       idents = "Skeletal")

nalx_foxo <- length(union(alx_foxo_1, alx_foxo_2))
```


The following code counts the number of co-expressing cells for Alx1 and FoxB (both high and low co-expression) in the He skeletogenic cell cluster.

```{r}
alx_foxb_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-11066.t1:Sp-FoxB` >= 0.5,
                       idents = "Skeletal")
alx_foxb_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-11066.t1:Sp-FoxB` >= 0.5,
                       idents = "Skeletal")

nalx_foxb <- length(union(alx_foxb_1, alx_foxb_2))
```


The following code counts the number of co-expressing cells for Alx1 and Tel (both high and low co-expression) in the He skeletogenic cell cluster.

```{r}
alx_tel_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12611.t1:Sp-Tel` >= 0.5,
                       idents = "Skeletal")
alx_tel_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 & 
                        `HER-12611.t1:Sp-Tel` >= 0.5,
                       idents = "Skeletal")

nalx_tel <- length(union(alx_tel_1, alx_tel_2))
```

This code calculates the proportions of Alx1 expressing cells in the skeletal lineage that express a larval skeletal or NSM marker, using the values obtained in the code blocks above.

```{r}
prop_scl <- nalx_scl/nalx # proportion of Alx1+ skeletogenic cells that co-express Scl
prop_ese <- nalx_ese/nalx # proportion of Alx1+ skeletogenic cells that co-express Ese
prop_gc <- nalx_gc/nalx # proportion of Alx1+ skeletogenic cells that co-express GataC
prop_tbr <- nalx_tbr/nalx # proportion of Alx1+ skeletogenic cells that co-express Tbr
prop_tel <- nalx_tel/nalx # proportion of Alx1+ skeletogenic cells that co-express Tel
prop_foxo <- nalx_foxo/nalx # proportion of Alx1+ skeletogenic cells that co-express FoxO
prop_foxb <- nalx_foxb/nalx # proportion of Alx1+ skeletogenic cells that co-express FoxB
```



## Co-expression analysis for Lv

### Generate co-expression plots for Figures 3 and 4

Obtain the UMAP coordinates for each of the cells in the Lv Seurat object

```{r}
lv_umap_coords <- as_tibble(Embeddings(lv.integrated, reduction = "umap"))
lv_umap_coords$Cell <- rownames(Embeddings(lv.integrated, reduction = "umap"))
```


Generate a co-expression plot for Alx1 and Scl for Lv.

```{r}
lv_scl05 <- WhichCells(lv.integrated,
                     expression = `LVA-36418.t1:Scl` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5)
lv_scl1 <- WhichCells(lv.integrated,
                     expression = `LVA-36418.t1:Scl` >= 1 &
                       `LVA-5802.t1:Alx1` >= 1)

coexp_plot(coords = lv_umap_coords, 
           low_exp = lv_scl05, 
           high_exp = lv_scl1)
```


Generate a co-expression plot for Alx1 and Ese for Lv.

```{r}
lv_ese05 <- WhichCells(lv.integrated,
                     expression = `LVA-5105.t1:Ese` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5)
lv_ese1 <- WhichCells(lv.integrated,
                     expression = `LVA-5105.t1:Ese` >= 1 &
                       `LVA-5802.t1:Alx1` >= 1)

coexp_plot(coords = lv_umap_coords, 
           low_exp = lv_ese05, 
           high_exp = lv_ese1)
```


Generate a co-expression plot for Alx1 and GataC for Lv.

```{r}
lv_gc05 <- WhichCells(lv.integrated,
                     expression = `LVA-32108.t1:GataC` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5)
lv_gc1 <- WhichCells(lv.integrated,
                     expression = `LVA-32108.t1:GataC` >= 1 &
                       `LVA-5802.t1:Alx1` >= 1)

coexp_plot(coords = lv_umap_coords, 
           low_exp = lv_gc05, 
           high_exp = lv_gc1)
```


Generate a co-expression plot for Alx1 and Tbr for Lv.

```{r}
lv_tbr05 <- WhichCells(lv.integrated,
                     expression = `LVA-12704.t1:Tbr` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5)
lv_tbr1 <- WhichCells(lv.integrated,
                     expression = `LVA-12704.t1:Tbr` >= 1 &
                       `LVA-5802.t1:Alx1` >= 1)

coexp_plot(coords = lv_umap_coords, 
           low_exp = lv_tbr05, 
           high_exp = lv_tbr1)
```


Generate a co-expression plot for Alx1 and Tel for Lv.

```{r}
lv_tel05 <- WhichCells(lv.integrated,
                     expression = `LVA-5116.t1:Tel` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5)
lv_tel1 <- WhichCells(lv.integrated,
                     expression = `LVA-5116.t1:Tel` >= 1 &
                       `LVA-5802.t1:Alx1` >= 1)

coexp_plot(coords = lv_umap_coords, 
           low_exp = lv_tel05, 
           high_exp = lv_tel1)
```


Generate a co-expression plot for Alx1 and FoxO for Lv.

```{r}
lv_foxo05 <- WhichCells(lv.integrated,
                     expression = `LVA-35307.t1:FoxO-1` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5)
lv_foxo1 <- WhichCells(lv.integrated,
                     expression = `LVA-35307.t1:FoxO-1` >= 1 &
                       `LVA-5802.t1:Alx1` >= 1)

coexp_plot(coords = lv_umap_coords, 
           low_exp = lv_foxo05, 
           high_exp = lv_foxo1)
```


Generate a co-expression plot for Alx1 and FoxB for Lv.

```{r}
lv_foxb05 <- WhichCells(lv.integrated,
                     expression = `LVA-21108.t1:FoxB` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5)
lv_foxb1 <- WhichCells(lv.integrated,
                     expression = `LVA-21108.t1:FoxB` >= 1 &
                       `LVA-5802.t1:Alx1` >= 1)

coexp_plot(coords = lv_umap_coords, 
           low_exp = lv_foxb05, 
           high_exp = lv_foxb1)
```


### Count numbers of coexpressing cells in the Lv skeletal lineage

The following code counts the total number of cells that express the Alx1 transcript in the Lv skeletogenic cell clusters.

```{r}
lvnalx <- length(WhichCells(lv.integrated,
                       expression = `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))
```


The following code counts the number of co-expressing cells for Alx1 and Scl, Ese, GataC, Tbr, Tel, FoxO, and FoxB (both high and low co-expression) in the Lv skeletogenic cell clusters.

```{r}
lvnalx_scl <- length(WhichCells(lv.integrated,
                     expression = `LVA-36418.t1:Scl` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))

lvnalx_ese <- length(WhichCells(lv.integrated,
                     expression = `LVA-5105.t1:Ese` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))

lvnalx_gc <- length(WhichCells(lv.integrated,
                     expression = `LVA-32108.t1:GataC` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))

lvnalx_tbr <- length(WhichCells(lv.integrated,
                     expression = `LVA-12704.t1:Tbr` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))

lvnalx_tel <- length(WhichCells(lv.integrated,
                     expression = `LVA-5116.t1:Tel` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))

lvnalx_foxo <- length(WhichCells(lv.integrated,
                     expression = `LVA-35307.t1:FoxO-1` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))

lvnalx_foxb <- length(WhichCells(lv.integrated,
                     expression = `LVA-21108.t1:FoxB` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres")))
```


This code calculates the proportions of Alx1 expressing cells in the skeletal lineage that express a larval skeletal or NSM marker, using the values obtained in the code blocks above.

```{r}
lvnalx_scl/lvnalx # proportion of Alx1+ skeletogenic cells that co-express Scl
lvnalx_ese/lvnalx # proportion of Alx1+ skeletogenic cells that co-express Ese
lvnalx_gc/lvnalx # proportion of Alx1+ skeletogenic cells that co-express GataC
lvnalx_tbr/lvnalx # proportion of Alx1+ skeletogenic cells that co-express Tbr
lvnalx_tel/lvnalx # proportion of Alx1+ skeletogenic cells that co-express Tel
lvnalx_foxo/lvnalx # proportion of Alx1+ skeletogenic cells that co-express FoxO
lvnalx_foxb/lvnalx # proportion of Alx1+ skeletogenic cells that co-express FoxB
```



## Analyze the number of Alx1 and Scl co-expressing cells over time

### Look at co-expression patterns in the He skeletogenic cell lineage

The following code identifies the Alx1 (both transcripts) expressing cells and the Alx1-Scl co-expressing cells in the He skeletogenic cell lineage.

```{r}
times <- str_extract(he.integrated@meta.data$Stage, "\\d+")
times <- as.integer(times)

he.integrated <- AddMetaData(he.integrated, times, col.name = "time_int")


alx_1 <- WhichCells(he.integrated, 
                       expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5,
                       idents = "Skeletal")
alx_2 <- WhichCells(he.integrated, 
                       expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5,
                       idents = "Skeletal")

he_alx <- union(alx_1, alx_2)


alx_scl_1 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.1.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-5916.t1:Sp-Scl` >= 0.5,
                       idents = "Skeletal")
alx_scl_2 <- WhichCells(he.integrated, 
                      expression = `HER-12794.t1.2.5f2c6949:Sp-Alx1` >= 0.5 &
                        `HER-5916.t1:Sp-Scl` >= 0.5,
                       idents = "Skeletal")

he_alx_scl <- union(alx_scl_1, alx_scl_2)
```

The cell metadata in the He Seurat object is filtered for the cells found in the Alx1 and Alx1-Scl expressing lists above, and the number of cells meeting each criterion is counted for each timepoint. This is plotted over time (see Figure 4E).

```{r}
he_alx_cells <- he.integrated@meta.data %>% 
  filter(rownames(he.integrated@meta.data) %in% he_alx) %>% 
  group_by(time_int) %>% 
  summarize(n = n()) %>% 
  mutate(Genes = "Alx1")

he_alx_scl_cells <- he.integrated@meta.data %>% 
  filter(rownames(he.integrated@meta.data) %in% he_alx_scl) %>% 
  group_by(time_int) %>% 
  summarize(n = n()) %>% 
  mutate(Genes = "Alx1 and Scl")

he_plot <- full_join(he_alx_cells, he_alx_scl_cells)

he_plot %>% 
  ggplot(mapping = aes(x = time_int, y = n, fill = Genes)) +
  geom_col(position = "dodge") +
  scale_x_continuous(breaks = c(12, 16, 20, 24, 30, 36, 42, 48, 54, 60)) +
  scale_fill_manual(values = c("Alx1" = "darkgray", "Alx1 and Scl" = "darkgreen"))+
  theme_bw() +
  labs(x = "Hours Post Fertilization",
       y = "Number of Cells")
```


### Look at co-expression patterns in the Lv skeletogenic cell lineage

The following code identifies the Alx1 (both transcripts) expressing cells and the Alx1-Scl co-expressing cells in the Lv skeletogenic cell lineage.

```{r}
times <- str_extract(lv.integrated@meta.data$Stage, "\\d+")
times <- as.integer(times)

lv.integrated <- AddMetaData(lv.integrated, times, col.name = "time_int")


lv_alx <- WhichCells(lv.integrated,
                       expression = `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres"))

lv_alx_scl <- WhichCells(lv.integrated,
                     expression = `LVA-36418.t1:Scl` >= 0.5 &
                       `LVA-5802.t1:Alx1` >= 0.5,
                     idents = c("PMCs", "Micromeres"))
```

The cell metadata in the Lv Seurat object is filtered for the cells found in the Alx1 and Alx1-Scl expressing lists above, and the number of cells meeting each criterion is counted for each timepoint. This is plotted over time (see Figure 4F).

```{r}
lv_alx_cells <- lv.integrated@meta.data %>% 
  filter(rownames(lv.integrated@meta.data) %in% lv_alx) %>% 
  group_by(time_int) %>% 
  summarize(n = n()) %>% 
  mutate(Genes = "Alx1")

lv_alx_scl_cells <- lv.integrated@meta.data %>% 
  filter(rownames(lv.integrated@meta.data) %in% lv_alx_scl) %>% 
  group_by(time_int) %>% 
  summarize(n = n()) %>% 
  mutate(Genes = "Alx1 and Scl")

lv_plot <- full_join(lv_alx_cells, lv_alx_scl_cells) %>% 
  add_row(time_int = 5, n = 0, Genes = "Alx1 and Scl")

lv_plot %>% 
  ggplot(mapping = aes(x = time_int, y = n, fill = Genes)) +
  geom_col(position = "dodge") +
  scale_x_continuous(breaks = c(5,6,7,8,9,10,11,12,13,14,15,16,18,20,24)) +
  scale_fill_manual(values = c("Alx1" = "darkgray", "Alx1 and Scl" = "darkgreen"))+
  theme_bw() +
  labs(x = "Hours Post Fertilization",
       y = "Number of Cells") +
  theme(legend.position = "none")
```


