---
title: "Neural Analysis"
author: "Brennan McDonald"
date: "2024-04-04"
output: html_document
---

# Overview

The following script provides the ability to run the analyses of the neural cell lineage in the He scRNA-seq dataset. The script contains the code needed to generate the plots used in Figure 6.

```{r}
# Load packages
library(tidyverse)
library(Seurat)
library(scCustomize)
```

```{r}
# Load the full He Seurat object
CreateSeuratObject(load(
  file = "HE_6-60hpf_integrated-SCT6k_195dim.Rda", 
                        verbose = T))
```


# Subsetting and re-clustering the neural cell cluster

Based on the marker genes expressed in the clusters of the full He Seurat object, I selected the clusters that showed expression of neural-related genes, such as Syt4 and Secrtag. I isolated these clusters into a new Seurat object, `neural_sub`.

```{r}
nerve_clusters <- c(31, 39, 51, 53, 54, 56)

neural <- subset(he.integrated, seurat_clusters %in% nerve_clusters)
```

I re-clustered the neural clusters at a higher resolution to learn about additional putative neural cell types. I then re-ran the UMAP algorithm on the re-clustered object to project the neural-only cells into UMAP space.

```{r}
# Recluster at higher resolution
neural_sub <- FindNeighbors(neural, dims = 1:195)
neural_sub <- FindClusters(neural_sub, resolution = 0.8) # 0.8 results in 17 clusters. This is quite robust, appears to be a consistent number

# re-run umap to give a neural-only projection
neural_sub <- RunUMAP(neural_sub, dims = 1:195)

# Save the neural Seurat object as a separate .Rda file
save(neural_sub, 
     file = "neural_subcluster-seurat-object.Rda")
```

Load the neural only Seurat object for downstream analyses.

```{r}
CreateSeuratObject(load(
  file = "neural_subcluster-seurat-object.Rda", 
                        verbose = T))
```


# Identifying neural cell types

I then used the FindAllMarkers function to identify the marker genes for the new neural clusters.

```{r}
# Get the marker genes for each cluster that are expressed in a minimum of 25% 
# of cells in the cluster and with a log2FC > 0.25
neural_sub.markers <- FindAllMarkers(neural_sub, only.pos = T, min.pct = 0.25, 
                                        logfc.threshold = 0.25)

export <- neural_sub.markers %>% 
  group_by(cluster) %>% 
  slice_max(n = 50, order_by = avg_log2FC)

write.csv(export, "data-files/top-neural-markers.csv",
          row.names = F)
```

I then used these marker genes, along with data from previous research on sea urchin nervous systems, to assign cell type identities to each of the neural clusters (see the marker genes used in Table S___).

```{r}
Idents(neural_sub) <- "seurat_clusters"

# The index of these celltype identities in the vector corresponds to the 
# number of the cluster that the identity will be assigned to
new.cluster.ids <- c("Differentiating Adult Neurons",
"Differentiating Adult Neurons",
"ANE Neural Progenitors",
"Adult Gut Neurons 1",
"Adult Endoderm Neuron Progenitors",
"Early ANE Neurons",
"Early ANE Neurons",
"Larval Enteric Neurons",
"Ciliary Band Neurons",
"Podia Neurons",
"Postoral Neurons",
"Adult Gut Neurons 2",
"Larval Serotonergic Neurons",
"Differentiated Neurons 1",
"ANE Light Receptor Neurons",
"Differentiated Neurons 2",
"Differentiated Neurons 3")

# The identities are added to a new identity class in the Seurat object
names(new.cluster.ids) <- levels(neural_sub)
neural_sub <- RenameIdents(neural_sub, new.cluster.ids)
Idents(neural_sub)

# The new identities are added to the metadata of the Seurat object
neural_sub <- AddMetaData(
  object = neural_sub,
  metadata = Idents(neural_sub),
  col.name = "celltype"
)

# Save the neural Seurat object with updated cell type annotations
save(neural_sub, 
     file = "neural_subcluster-seurat-object.Rda")
```


Generate the UMAP DimPlots of the neural cells colored by sample time point and cell type identity.

```{r}
# Generate a list of sample timepoints (the number values only)
times <- str_extract(neural_sub@meta.data$Stage, "\\d+")
times <- as.integer(times)

neural_sub <- AddMetaData(neural_sub, times, col.name = "time_int")

x <- seq(10,250,20)
viridis_plasma_dark_high[x]

# This generates the UMAP colored by time used in Figure 6
DimPlot_scCustom(neural_sub,
                 group.by = "time_int",
                 colors_use = viridis_plasma_dark_high[x]) +
  theme(legend.position = "none")

# This generates the UMAP colored by celltype used in Figure 6
DimPlot_scCustom(neural_sub,
                 group.by = "celltype",
                 colors_use = "alphabet") +
  theme(legend.position = "none")
```


The following code uses the curated set of neural cell type marker genes (from Table S___) to generate a dotplot showing expression of the genes in each cell type for Figure S___.

```{r}
neural_celltype_markers <- c("HER-3624.t1:Sp-SoxC",
"HER-10974.t1:Sp-Hbn",
"HER-26992.t1:Sp-Acsc",
"HER-17866.t1:Sp-Six3",
"HER-45795.t1:Sp-Secrtag",
"HER-10174.t1:Sp-Syt4",
"HER-17538.t1:Sp-CalmL1",
"HER-37930.t1:Sp-Z133-1",
"HER-12508.t1-HER-12507.t1:Sp-Gpr51-4",
"HER-10076.t1:Sp-Gpr103",
"HER-9250.t2:Sp-Ebf",
"HER-16478.t1:Sp-Drd1L-1",
"HER-45819.t1:Sp-Z116",
"HER-10349.t1:Sp-Isl",
"HER-30955.t1:Sp-NeuroD",
"HER-12505.t1.1.5f2c6944:Sp-Pax6-1",
"HER-6581.t1:Sp-FoxD-1",
"HER-45737.t1-HER-45739.t1:Sp-Opsin3.1",
"HER-43627.t1:Sp-Ets1/2",
"HER-2616.t1:Sp-Nos-1",
"HER-10476.t1:Sp-Otp",
"HER-26132.t1:Sp-Th",
"HER-33009.t1:Sp-Onecut",
"HER-46843.t1:Sp-Dbh",
"HER-3994.t1:Sp-Gpr133-2",
"HER-34974.t1:Sp-Brn1/2/4",
"HER-23959.t1:Sp-Chat",
"HER-20.t1:Sp-Sert",
"HER-27402.t1:Sp-Tph",
"HER-10374.t1-HER-10375.t1:Sp-Gpr54-1",
"HER-24945.t1:Sp-Opsin2",
"HER-17315.t1:Sp-Opsin1",
"HER-2641.t1:Sp-Lim1",
"HER-9123.t1:Sp-Cub/Ldla/Lrr/Gpcr",
"HER-48798.t1-HER-48799.t1:Sp-Phox2",
"HER-39663.t1-HER-39664.t1:Sp-Hb9",
"HER-44934.t1:Sp-Scratch",
"HER-8788.t1:Sp-Hmx",
"HER-22681.t1:Sp-GrprL",
"HER-17257.t1:Sp-Gpr51",
"HER-31310.t1:Sp-Bursicona",
"HER-36590.t1:Sp-Rxfp2-6",
"HER-24113.t1-HER-24114.t1:Sp-Bmp3",
"HER-16539.t1:Sp-Sstr4L",
"HER-11091.t1:Sp-Adra1aL-4",
"HER-15597.t1:Sp-Unc4.1-1",
"HER-24299.t1:Sp-Trpc5-2",
"HER-24249.t1:Sp-Arxl")

levs <- factor(c("ANE Neural Progenitors",
                 "Differentiating Adult Neurons",
                 "Early ANE Neurons",
                 "Larval Serotonergic Neurons",
                 "Postoral Neurons",
                 "Ciliary Band Neurons",
                 "ANE Light Receptor Neurons",
                 "Larval Enteric Neurons",
                 "Adult Endoderm Neuron Progenitors",
                 "Podia Neurons",
                 "Adult Gut Neurons 1",
                 "Adult Gut Neurons 2",
                 "Differentiated Neurons 1",
                 "Differentiated Neurons 2",
                 "Differentiated Neurons 3"))

Idents(neural_sub) <- factor(neural_sub@active.ident, levels = levs)

# This generates the dotplot of neural cell type marker genes for Figure S___
# Export at 8.5x11
DotPlot_scCustom(neural_sub,
                 features = neural_celltype_markers,
                 colors_use = viridis_dark_high,
                 x_lab_rotate = T,
                 flip_axes = T)
```


# Get neural TF list

The following section provides the code used to generate the clustered DotPlot of neural transcription factor gene expression in Figure 6.

I first began by compiling a list of neural transcription factors published in Burke et al. (2006).

```{r}
hery_spu <- read_tsv("data-files/H_ery_SPU_hits_best.txt")
neural_spu <- read_tsv("data-files/neural_tfs.txt")
```


I then merged the list of transcription factor gene names with the gene annotations for He from Davidson et al. (2022). I also added a few transcription factors that had been shown in other publications to have neural expression patterns.

```{r}
neural_hery <- hery_spu %>% 
  filter(SPU.hit %in% neural_spu$SPU)

add <- hery_spu %>% 
  filter(SPU.common_name %in% c("Sp-Scratch", "Sp-Hmx",
                                "Sp-Z133_1", "Sp-HesD", "Sp-HesA",
                                "Sp-Z166"))

neural_hery <- bind_rows(neural_hery, add)

# Add an additional column to the dataframe with gene names that correspond to 
# the names in the Seurat object.
neural_tf <- neural_tf %>% 
  mutate(new_hery_id = str_replace_all(H_ery_ID, "_", "-"),
         new_spu = str_replace_all(SPU.common_name, "_", "-")) %>% 
  mutate(seurat_code = paste(new_hery_id, new_spu, sep = ":")) %>% 
  select(-new_hery_id, -new_spu)

write.csv(neural_hery, "data-files/neural_tfs_hery.csv",
          row.names = F)

# also removed FoxL2, Rfx3, which were not found in the seurat object
# Removed non-expressed transcripts
# Removed HesB, HesC
```


I then loaded the full He Seurat object and identified the neural transcription factors that had the highest average expression across all cells with a neural identity. The top 30 expressed genes were plotted in the clustered DotPlot presented in Figure 6.

```{r}
neural_tf <- read_csv("data-files/neural_tfs_hery.csv")


# Gets the average expression for genes in all cell types
av_exp <- AverageExpression(he.integrated)

# Selects only genes with a neural cell type identity
neural_exp <- av_exp$RNA[,"Neural"]

# The genes are ordered by expression level
neural_exp <- tibble(gene = names(neural_exp), expression = neural_exp) %>% 
  arrange(desc(expression))

# This list is filtered only for transcription factors found in my curated list.
exp_neural_tf <- neural_exp %>% 
  filter(gene %in% neural_tf$seurat_code) %>% 
  mutate(gene_name = str_split_i(gene, ":", 2))

# The top 30 expressed neural transcription factors are plotted in the 
# clustered DotPlot for Figure 6.
top30_clusterplot <- Clustered_DotPlot(seurat_object = neural_sub,
                                 features = exp_neural_tf$gene[1:30],
                                 k = 10,
                                 column_label_size = 0,
                                 seed = 1,
                                 colors_use_idents = 
                                   DiscretePalette_scCustomize(15,palette = "alphabet"),
                                 colors_use_exp = viridis_dark_high)

top30_clusterplot[[2]]
```





